keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -693,8 +1310,188 @@ type NamedResourcesStringSlice struct {
 All names must be DNS sub-domains. This excludes the "/" character, therefore
 combining different names with that separator to form an ID is valid.
 
+In the Go types above, all structs starting with `NamedResources` are part of
+that structured model. Code generators (more specifically, the applyconfig
+generator) assume that all Go types of an API are defined in the same Go
+package. If it wasn't for that, defining those structs in their own package
+without the `NamedResources` prefix would be possible and make the Go code
+cleaner without affecting the Kubernetes API.
 
-### ResourceClaimParameters
+##### ResourceClass
+
+```go
+// ResourceClass is used by administrators to influence how resources
+// are allocated.
+//
+// This is an alpha type and requires enabling the DynamicResourceAllocation
+// feature gate.
+type ResourceClass struct {
+    metav1.TypeMeta
+    // Standard object metadata
+    // +optional
+    metav1.ObjectMeta
+
+    // ParametersRef references an arbitrary separate object that may hold
+    // parameters that will be used by the driver when allocating a
+    // resource that uses this class. A dynamic resource driver can
+    // distinguish between parameters stored here and and those stored in
+    // ResourceClaimSpec.
+    // +optional
+    ParametersRef *ResourceClassParametersReference
+
+    // Only nodes matching the selector will be considered by the scheduler
+    // when trying to find a Node that fits a Pod when that Pod uses
+    // a ResourceClaim that has not been allocated yet.
+    //
+    // Setting this field is optional. If null, all nodes are candidates.
+    // +optional
+    SuitableNodes *core.NodeSelector
+}
```

Comment:
We can also make `ResourceClaim.ResourceClassName` optional and add a default, like literally `default`. The users don't need to bother with figuring out which classes might exist in any given cluster and admins who want to influence allocation can do it reliably.

The implication for DRA drivers is that they don't need to provide a ResourceClass as part of their install instructions.

I actually like this.

Reply:
What is the difference between your first and second comment? In the first you say it should be mandatory, and in the second you say it should be optional. Are you talking about different a different `ResourceClaim` in each comment (i.e. the actual object vs. the field in the pod spec)?

Reply:
Same field (`ResourceClaim.ResourceClassName`), different alternatives: either keeping it mandatory or making it optional with a default.

Reply:
If there is heterogeneity in the cluster, isn't the case admins have to use ReseourceClassName?

Reply:
Admins create ResourceClasses, users set `ResourceClaim.ResourceClassName` based on information provided to them by the admin that tells them which class to use for which purpose. This wouldn't go away, I just don't see why it needs to be required.

Reply:
I don't quite follow this. Without a `ResourceClassName`, how does the scheduler know which driver is responsible for the resource being requested (so it can look at the appropriate resource slice to allocate from).

I don't see how we can go from the following claim (which is valid today):
```
```
apiVersion: resource.k8s.io/v1alpha2
kind: ResourceClaim
metadata:
  name: my-gpu
spec:
  resourceClassName: gpu.nvidia.com
```

to this one and expect it to work:

```
apiVersion: resource.k8s.io/v1alpha2
kind: ResourceClaim
metadata:
  name: my-gpu
  ```

Reply:
```
apiVersion: resource.k8s.io/v1alpha2
kind: ResourceClaim
metadata:
  name: my-gpu
spec:
  resourceClassName: gpu.nvidia.com
```

This currently works because we [generate default claim parameters](https://github.com/kubernetes/kubernetes/blob/17854f0e0a153b06f9d0db096e2cd8ab2fa89c11/pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go#L1019-L1043) with the `ResourceClass.DriverName` embedded inside it.

This was already slightly hackish and https://github.com/kubernetes/kubernetes/pull/124251 is replacing those generated parameters with something defined inside the class. I wanted to document that in the KEP in a follow-up PR - let's see whether we need it now.

/cc @ravisantoshgudimetla 

With the current content in this KEP, your second "empty" claim basically doesn't do anything. It's not invalid, it simply doesn't request any resources. Such a claim can be referenced, we can allocate it with zero resource handles, and kubelet doesn't need to do anything for it. I'm unsure whether validation should reject it. When the content of a ResourceClaim is generated, "request nothing" becomes a valid corner case. Handling that in Kubernetes instead of the code which generates the ResourceClaim may be useful.

That the claim doesn't do anything shouldn't be a surprise to the user, because it's obvious that it's empty.

If a user wants a claim with no explicit resource requests to do something, they have to reference a resource class and that resource class has to reference default parameters because `ResourceClass.DriverName` is no longer part of KEP #4381. So we do need the default parameters already now. In contrast to the "empty claim", I would treat a claim with a `ResourceClassName` of a class with no default parameters as an error and refuse to allocate the claim because the users was expecting to get something.

Reply:
In the above sample, the first claim allocates the entire GPU. In the second example `ResourceClassName` is taken from the default value set by the admin. From the UX standpoint, is there a way for users to discover the default `ResourceClassName`?

Reply:
So instead of *requiring* a `ResourceClassName`, your suggesting to do anything useful we either need to:
1. Include an explicit `ResourceClaimParameters` (which can then be used to figure out which driver is in charge of providing resources for the claim), or;
2. Actually provide a `ResourceClassName`, where the admin will now be required to set a default `ResourceClaimParameters` (in lieu of setting a required `DriverName` field).

Reply:
@asm582: in the first example it is not clear what using "gpu.nvidia.com" does. It depends on the defaults associated with that class.

If we decide to make `ResourceClassName` optional with a way to have a default class, then there several options:
- Fixed name for the default resource class ("default" ?).
- Use a field or annotation to mark one of many classes as the default, similar to [storage classes](https://kubernetes.io/docs/tasks/administer-cluster/change-default-storage-class/#changing-the-default-storageclass).

This storage class API with annotation is unusual, a field is probably preferred. I don't know whether making a storage class default configurable like that is actually considered a good UX - probably worth asking the SIG Storage folks.

@klueska: basically yes. Another variation would be to allow "ResourceClassName set, no default parameters" and treat it like the empty claim.

Reply:
Let's keep this simple:
- ResourceClassName may be empty.
- If no parameters are provided (neither in claim nor class), nothing needs to be allocated. Same for empty claim.
- No default ResourceClass.

I updated the text with this proposal: https://github.com/kubernetes/enhancements/compare/4fe1b58595ea013e08a4ede9930e7a5d244e9e68..d028f775ae50f60c66bfdcb7b28a8aedb4a6563b

Reply:
I think this is fine here. But I think we need to iterate on these kinds of decisions as we build out the prototype. The results of that will eventually land in here.

Reply:
@klueska: Are you also okay with the ResourceClassName update above?

@johnbelamaric: I add an UNRESOLVED to the API section. It includes some of the examples of things that we might want to change, but doesn't need to be complete - the important part is to have that marker.

Let's LGTM and approve this update so that we can move on.